raster-foundry/granary
# Add Batch Prediction Infrastructure

## Overview

These changes were made in an effort to add Batch prediction infrastructure:

- Add GPU-based Batch compute environment, job queue, and example job definition
- Add IAM resources for Batch container instances and Spot Fleet

Closes #15 

## Testing Instructions

- Execute Terraform within the appropriate container and trigger a plan/apply cycle with the specified `GIT_COMMIT` override

```console
$ docker-compose -f docker-compose.ci.yml run --rm terraform
bash-5.0# GIT_COMMIT="2e91448" ./scripts/infra plan
bash-5.0# GIT_COMMIT="2e91448" ./scripts/infra apply
```

- Submit the example job definition

```bash
$ aws batch submit-job \
     --job-name "test-gpu" \
     --job-queue queueGranaryGPU \
     --job-definition jobGranaryTestGPU
{
    "jobName": "test-gpu",
    "jobId": "0a8785a2-8c9f-4a5c-a580-3e784e7549a9"
}
```

- Confirm the job execution is successful and the output of `nvidia-smi` shows correct GPU device state

```bash
$ aws batch describe-jobs \
    --jobs "0a8785a2-8c9f-4a5c-a580-3e784e7549a9" | jq -r .jobs[0].container.logStreamName
test_gpu_job_definition/default/ea6caeb0-ddea-44bc-abca-31025616ed92
```

```bash
$ aws logs get-log-events \
    --log-group-name /aws/batch/job \
    --log-stream-name "jobGranaryTestGPU/default/ea6caeb0-ddea-44bc-abca-31025616ed92" | jq -r .[][].message
Thu Jan  2 18:36:44 2020
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.87.00    Driver Version: 418.87.00    CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  On   | 00000000:00:1E.0 Off |                    0 |
| N/A   42C    P0    28W / 300W |      0MiB / 16130MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
. . .
```
